### PR TITLE
Skip Setting Default Log Shipping Monitor

### LINKED ISSUE #8394: Invoke-DbaDBLogShipping unsuccessfully completes with -Force flag, restore operation fails
### Verified issue does not already exist?

Yes

### What error did you receive?

While running the restore job on the destination server:

```
Message
Microsoft (R) SQL Server Log Shipping Agent
[Assembly Version = 15.0.0.0, File Version = 15.0.4223.1]
Copyright (c) 2016 Microsoft. All rights reserved.

2022-06-09 12:21:30.92	----- START OF TRANSACTION LOG RESTORE   -----
2022-06-09 12:21:31.06	*** Error: Could not retrieve restore settings.(Microsoft.SqlServer.Management.LogShipping) ***
2022-06-09 12:21:31.06	*** Error: The specified agent_id F363212F-E96B-4988-9AC2-98653D7D4CA9 or agent_type 2 do not form a valid pair for log shipping monitoring processing.(.Net SqlClient Data Provider) ***
2022-06-09 12:21:31.06	*** Error: Could not log history/error message.(Microsoft.SqlServer.Management.LogShipping) ***
2022-06-09 12:21:31.06	*** Error: The specified agent_id F363212F-E96B-4988-9AC2-98653D7D4CA9 or agent_type 2 do not form a valid pair for log shipping monitoring processing.(.Net SqlClient Data Provider) ***
2022-06-09 12:21:31.06	*** Error: Could not cleanup history.(Microsoft.SqlServer.Management.LogShipping) ***
2022-06-09 12:21:31.06	*** Error: The specified agent_id F363212F-E96B-4988-9AC2-98653D7D4CA9 or agent_type 2 do not form a valid pair for log shipping monitoring processing.(.Net SqlClient Data Provider) ***
2022-06-09 12:21:31.06	----- END OF TRANSACTION LOG RESTORE     -----
```

### Steps to Reproduce

```powershell
$params = @{
    SourceSqlInstance = 'source-server.domain.com'
    DestinationSqlInstance = 'destination-server.domain.com'
    Database = 'migrationDB'
    BackupNetworkPath = '\\networkshare\ls'
    CopyDestinationFolder = '\\networkshare\ls'
    GenerateFullBackup = $true
    CompressBackup = $true
    SecondaryDatabaseSuffix = '_ls'
    RestoreDataFolder = 'R:\path\data\'
    RestoreLogFolder = 'R:\path\log\'
    BackupSchedule='Logshipping Backup'
    BackupScheduleFrequencyType = 'daily'
    BackupScheduleFrequencyInterval = 1
    BackupScheduleFrequencySubdayType = 'minutes'
    BackupScheduleFrequencySubdayInterval = 15
    CopyScheduleDisabled = $true
    RestoreScheduleFrequencyType = 'daily'
    RestoreScheduleFrequencyInterval = 1
    RestoreScheduleFrequencySubdayType = 'minutes'
    RestoreScheduleFrequencySubdayInterval = 5
    }
```
The first invocation **fails**:
`Invoke-DbaDbLogShipping @params -Force`

Clear out log shipping in between invocations:
`Remove-DbaDbLogShipping -PrimarySqlInstance source-server.domain.com -SecondarySqlInstance destination-server.domain.com -Database migrationDB`

Second invocation is **successful**:
`Invoke-DbaDbLogShipping @params`

### Are you running the latest release?

Yes

### Other details or mentions

Both exit with the same message:
```
PrimaryInstance   : source-server.domain.com
SecondaryInstance : destination-server.domain.com
PrimaryDatabase   : migrationDB
SecondaryDatabase : migrationDB_ls
Result            : Success
Comment           :
```

When I query the primary after invoking **with** the `-Force` flag, the log shipping monitoring tables are populated as **expected**.
```
EXEC master.dbo.sp_help_log_shipping_monitor
select * from msdb.dbo.log_shipping_monitor_primary
```
image

When I query the secondary after invoking **with** the `-Force` flag, the log shipping monitoring tables are empty, this is **unexpected**.
```
EXEC master.dbo.sp_help_log_shipping_monitor
select last_restored_date, * from msdb.dbo.log_shipping_monitor_secondary
```
image

---

When I query the primary after invoking **without** the `-Force` flag, the log shipping monitoring tables are populates as **expected**.
image

When I query the secondary after invoking **without** the `-Force` flag, the log shipping monitoring tables are populates as **expected**.
image

---

My first reaction is that there is something failing within the `New-dbaLogShippingSecondaryPrimary` internal function. I haven't been able to debug it successfully.

### What PowerShell host was used when producing this error

PowerShell Core (pwsh.exe)

### PowerShell Host Version

Name                           Value
----                           -----
PSVersion                      7.2.3
PSEdition                      Core
GitCommitId                    7.2.3
OS                             Microsoft Windows 6.3.9600
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0.}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

### SQL Server Edition and Build number

source-server.domain.com: 
```
Microsoft SQL Server 2017 (RTM-CU22) (KB4577467) - 14.0.3356.20 (X64)   Aug 20 2020 22:33:27   Copyright (C) 2017 Microsoft Corporation  Standard Edition (64-bit) on Windows Server 2012 R2 Standard 6.3 <X64> (Build 9600: ) (Hypervisor) 
```

destination-server.domain.com:
```
Microsoft SQL Server 2019 (RTM-CU16) (KB5011644) - 15.0.4223.1 (X64)   Apr 11 2022 16:24:07   Copyright (C) 2019 Microsoft Corporation  Enterprise Evaluation Edition (64-bit) on Windows Server 2019 Datacenter 10.0 <X64> (Build 17763: ) (Hypervisor) 
```

### .NET Framework Version

.NET 6.0.4

### LINKED ISSUE #8670: Invoke-DbaDbLogShipping not updating Secondary server fully
### Verified issue does not already exist?

I have searched and found no existing issue

### What error did you receive?

No error is generated by the Invoke-DbaDbLogShipping cmdlet, it indicates Successful operations. The cmdlet does indeed create the Back, Copy and Restore jobs and restores the database to the secondary server, but when you try to open the Secondary server properties for the log shipping in SSMS, it throws an error "Management Studio could not retrieve information about secondary database". Checking the status of the Copy job on the secondary server, errors such as the following can be found.

Message
2022-11-30 14:40:03.00	*** Error: Could not retrieve copy settings for secondary ID '36b404be-6296-4f30-950d-9fbb467d7acd'.(Microsoft.SqlServer.Management.LogShipping) ***
2022-11-30 14:40:03.00	*** Error: The specified agent_id 36B404BE-6296-4F30-950D-9FBB467D7ACD or agent_type 1 do not form a valid pair for log shipping monitoring processing.(.Net SqlClient Data Provider) ***
2022-11-30 14:40:03.00	*** Error: Could not log history/error message.(Microsoft.SqlServer.Management.LogShipping) ***
2022-11-30 14:40:03.00	*** Error: The specified agent_id 36B404BE-6296-4F30-950D-9FBB467D7ACD or agent_type 1 do not form a valid pair for log shipping monitoring processing.(.Net SqlClient Data Provider) ***
2022-11-30 14:40:03.00	*** Error: Could not cleanup history.(Microsoft.SqlServer.Management.LogShipping) ***
2022-11-30 14:40:03.00	*** Error: The specified agent_id 36B404BE-6296-4F30-950D-9FBB467D7ACD or agent_type 1 do not form a valid pair for log shipping monitoring processing.(.Net SqlClient Data Provider) ***
2022-11-30 14:40:03.00	----- END OF TRANSACTION LOG COPY     -----

Finally, when checking msdb.dbo.log_shipping_monitor_secondary, we find no record of the secondary database listed. This is likely the root cause of the error in SSMS and why the Copy job would fail.

### Steps to Reproduce

```powershell
# provide your command(s) executed pertaining to dbatools
# please include variable values (redacted or fake if needed) for reference

$params = @{
SourceSqlInstance = 'Server1'
DestinationSqlInstance = 'Server2'
Database = 'TestDatabase'
BackupScheduleFrequencyType = 'daily'
BackupScheduleFrequencyInterval = 1
BackupScheduleFrequencySubdayType = 'Minutes'
BackupScheduleFrequencySubdayInterval = 5
CompressBackup = $true
SharedPath= '\\Server1\LogShippingLogFiles$\Source'
LocalPath= 'X:\Backup\LogShippingLogFiles\Source'
CopyScheduleFrequencyType = 'daily'
CopyScheduleFrequencyInterval = 1
CopyScheduleFrequencySubdayType = 'Minutes'
CopyScheduleFrequencySubdayInterval = 5
CopyDestinationFolder = '\\Server2\LogShippingLogFiles$\Destination'
GenerateFullBackup = $true
RestoreScheduleFrequencyType = 'daily'
RestoreScheduleFrequencyInterval = 1
RestoreScheduleFrequencySubdayType = 'Hours'
RestoreScheduleFrequencySubdayInterval = 4
RestoreAlertThreshold = 300
Force = $true
}
Invoke-DbaDbLogShipping @params
```

- Open database properties and try to check the properties of the secondary server. Result: Error is thrown.

- Check Status of Copy job on secondary server, Errors are logged.

- Run the following on the secondary server. Result: No record of LS database.
```sql
SELECT * FROM msdb.dbo.log_shipping_monitor_secondary
GO
```

### Please confirm that you are running the most recent version of dbatools

Version
----------
1.1.143

### Other details or mentions

_No response_

### What PowerShell host was used when producing this error

Windows PowerShell (powershell.exe)

### PowerShell Host Version

Name                           Value
----                           -----
PSVersion                      5.1.19041.1682
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.19041.1682
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

### SQL Server Edition and Build number

SQL Server 2016 Standard 13.0.7016.1

### .NET Framework Version

PSChildName                      Version
-----------                      -------
v2.0.50727                       2.0.50727.4927
v3.0                             3.0.30729.4926
Windows Communication Foundation 3.0.4506.4926
Windows Presentation Foundation  3.0.6920.4902
v3.5                             3.5.30729.4926
Client                           4.8.04084
Full                             4.8.04084
Client                           4.0.0.0

### PR DESCRIPTION
**Pull request by sqldeployhelmet**
## Please read -- recent changes to our repo
On November 10, 2022, [we removed some bloat from our repository (for the second and final time)](https://github.com/dataplat/dbatools/issues/8542). This change requires that all contributors reclone or refork their repo.

PRs from repos that have not been recently reforked or recloned will be closed and @potatoqualitee will cherry-pick your commits and open a new PR with your changes.

 - [X] Please confirm you have the smaller repo (85MB .git directory vs 275MB or 110MB or 185MB .git directory)

## Type of Change

 - [X] Bug fix (non-breaking change, Possibly fixes #8670, fixes #8394 )
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (affects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/dataplat/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

### Purpose

The current behavior of dbatools is to default to setting the primary instance as the monitoring instance. When passing in the primary server as a default it is possible encounter a kerberos delegation error (double-hop). In my case this presented as an un-handled "non-success" when dbatools ran `sp_add_log_shipping_secondary_primary`. The errors I experienced seem similar to issues #8670 and #8394 and this may prove to be a fix for them.

Providing a monitor server is optional if the monitored server is not passed to configuration stored procedures SQL will create the LS alert jobs on the instance where the stored procedure is run. and can cause issues configuring log shipping. The added "non-success" handling will also now alert the user if an unexpected result was returned when executing log shipping configuration stored procedures.

The fix provided was testing with SQL 2019 in my limited work test environment. The little code that was specific to earlier editions (anything earlier than SQL 2014) was left un-touched at this time.
### Approach

This update removes the defaulting to the primary instance as the monitor server. This then excludes setting the monitor server parameter when calling the log shipping set up stored procecures. SQL then defaults to creating the appropriate LS alert job local to the primary or secondary database(s).

Additionally SQL was added when executing log shipping configuration stored procedures in case they do not succeed and do not throw an error. The unexpected return code (any
 non-zero result) is caught and raised as an error to dbatools.

### Commands to test

* `Invoke-DbaDbLogshipping`
  *  `New-DbaLogShippingPrimaryDatabase`
  *  `New-DbaLogShippingSecondaryPrimary`
  *  `New-DbaLogShippingSecondaryDatabase`

### CODE REVIEW

File: private/functions/New-DbaLogShippingPrimaryDatabase.ps1 [hunk 1]
```diff
@@ -134,7 +134,7 @@ function New-DbaLogShippingPrimaryDatabase {
         Stop-Function -Message "The backup share path $BackupShare should be formatted in the form \\server\share." -Target $SqlInstance
         return
     } else {
-        if (-not ((Test-Path $BackupShare -PathType Container -IsValid) -and ((Get-Item $BackupShare).PSProvider.Name -eq 'FileSystem'))) {
+        if (-not ((Test-DbaPath $BackupShare -SqlInstance $SqlInstance -SqlCredential $SqlCredential) -and ((Get-Item $BackupShare).PSProvider.Name -eq 'FileSystem'))) {
```
Comment (wsmelton) [inline private/functions/New-DbaLogShippingPrimaryDatabase.ps1:137]:
```suggestion
        if (-not ((Test-DbaPath $BackupShare -SqlInstance $server) -and ((Get-Item $BackupShare).PSProvider.Name -eq 'FileSystem'))) {
```

File: private/functions/New-DbaLogShippingSecondaryDatabase.ps1 [hunk 2]
```diff
@@ -237,6 +237,15 @@ function New-DbaLogShippingSecondaryDatabase {
         $Query += ";"
     }
 
+    $Query += "
+        IF (@SP_Add_RetCode <> 0)
```
Comment (wsmelton) [inline private/functions/New-DbaLogShippingSecondaryDatabase.ps1:241]:
```suggestion
        IF (@SP_Add_RetCode <> AND @@ERROR <> 0)
```

Is it worth also checking for the latest error?

File: private/functions/New-DbaLogShippingSecondaryPrimary.ps1 [hunk 3]
```diff
@@ -131,7 +132,7 @@ function New-DbaLogShippingSecondaryPrimary {
         Stop-Function -Message "The backup destination path should be formatted in the form \\server\share." -Target $SqlInstance
         return
     } else {
-        if (-not ((Test-Path $BackupDestinationDirectory -PathType Container -IsValid) -and ((Get-Item $BackupDestinationDirectory).PSProvider.Name -eq 'FileSystem'))) {
+        if (-not ((Test-DbaPath $BackupDestinationDirectory -SqlInstance $SqlInstance -SqlCredential $SqlCredential) -and ((Get-Item $BackupDestinationDirectory).PSProvider.Name -eq 'FileSystem'))) {
```
Comment (wsmelton) [inline private/functions/New-DbaLogShippingSecondaryPrimary.ps1:135]:
```suggestion
        if (-not ((Test-DbaPath $BackupDestinationDirectory -SqlInstance $ServerSecondary) -and ((Get-Item $BackupDestinationDirectory).PSProvider.Name -eq 'FileSystem'))) {
```

Please reuse connections already created via Connect call. This also makes it clear we are checking the path on the secondary server.

File: private/functions/New-DbaLogShippingSecondaryPrimary.ps1 [hunk 4]
```diff
@@ -203,6 +204,15 @@ function New-DbaLogShippingSecondaryPrimary {
         $Query += ";"
     }
 
+    # catch any non-success non error and throw:
+    $Query += "
+        IF (@SP_Add_RetCode <> 0)
```
Comment (wsmelton) [inline private/functions/New-DbaLogShippingSecondaryPrimary.ps1:209]:
```suggestion
        IF (@SP_Add_RetCode <> 0 AND @@ERROR = 0)
```

Is it worth also checking for last error code?

File: public/Invoke-DbaDbLogShipping.ps1 [hunk 5]
```diff
@@ -212,7 +212,7 @@ function Invoke-DbaDbLogShipping {
 
     .PARAMETER PrimaryMonitorServer
         Is the name of the monitor server for the primary server.
-        The default is the name of the primary sql server.
+        If ommitted logshipping will be monitored locally on the primary server.
```
Comment (wsmelton) [inline public/Invoke-DbaDbLogShipping.ps1:215]:
```suggestion
        Defaults to monitor on the instance provided via SourceSqlInstance param.
```

File: public/Invoke-DbaDbLogShipping.ps1 [hunk 6]
```diff
@@ -308,7 +308,7 @@ function Invoke-DbaDbLogShipping {
 
     .PARAMETER SecondaryMonitorServer
         Is the name of the monitor server for the secondary server.
-        The default is the name of the secondary sql server.
+        If ommitted logshipping will be monitored locally on the secondary server.
```
Comment (wsmelton) [inline public/Invoke-DbaDbLogShipping.ps1:311]:
```suggestion
        Defaults to monitor on the instance provided via DestinationSqlInstance param.
```

File: public/Invoke-DbaDbLogShipping.ps1 [hunk 7]
```diff
@@ -1376,12 +1376,6 @@ function Invoke-DbaDbLogShipping {
                     }
                 }
 
-                # Check the primary monitor server
-                if ($Force -and (-not $PrimaryMonitorServer -or [string]$PrimaryMonitorServer -eq '' -or $null -eq $PrimaryMonitorServer)) {
-                    Write-Message -Message "Setting monitor server for primary server to $SourceSqlInstance." -Level Verbose
-                    $PrimaryMonitorServer = $SourceSqlInstance
-                }
-
```
Comment (wsmelton) [inline public/Invoke-DbaDbLogShipping.ps1:1384]: This is breaking change is it not?

File: public/Invoke-DbaDbLogShipping.ps1 [hunk 8]
```diff
@@ -1404,12 +1398,6 @@ function Invoke-DbaDbLogShipping {
                     }
                 }
 
-                # Check the secondary monitor server
-                if ($Force -and (-not $SecondaryMonitorServer -or [string]$SecondaryMonitorServer -eq '' -or $null -eq $SecondaryMonitorServer)) {
-                    Write-Message -Message "Setting secondary monitor server for $destInstance to $SourceSqlInstance." -Level Verbose
-                    $SecondaryMonitorServer = $SourceSqlInstance
-                }
-
```
Comment (wsmelton) [inline public/Invoke-DbaDbLogShipping.ps1:1412]: This is a breaking change is it not?

**Review by wsmelton — CHANGES_REQUESTED** [review]

File: public/Invoke-DbaDbLogShipping.ps1 (hunk 7 above)
Comment (potatoqualitee) [reply public/Invoke-DbaDbLogShipping.ps1:1384]: Looks like it but this is a good time, as we're about to release a new version.

**Comment by potatoqualitee** [comment]
I shall accept this breaking change, thank you 🙇🏼  /cc @sanderstad

**Pull request merged**